azure-pipelines.yml

```diff
@@ -9,20 +19,60 @@ trigger:
     include:
       - main
 
+variables:
+  # If the user didn't override the signing type, then only real-sign on main.
+  ${{ if ne(parameters.SignTypeOverride, 'default') }}:
+    SignType: ${{ parameters.SignTypeOverride }}
+  ${{ if and(eq(parameters.SignTypeOverride, 'default'), or(startsWith(variables['Build.SourceBranch'], 'refs/tags'), eq(variables['Build.SourceBranchName'], 'main'))) }}:
+    SignType: real
+  ${{ if and(eq(parameters.SignTypeOverride, 'default'), not(or(startsWith(variables['Build.SourceBranch'], 'refs/tags'), eq(variables['Build.SourceBranchName'], 'main')))) }}:
+    SignType: test
+
 jobs:
-  - job: BuildPublish
-    displayName: Build and publish
+  - template: jobs/shared/build.yml
+  - job: GetNpmPackageVersions
+    dependsOn: Build
+    displayName: Extract npm package versions into variables
+    steps:
+      - script: echo "##vso[task.setvariable variable=CurrentVersion;isoutput=true]$(npm view @microsoft/vscode-serial-monitor-api version)"
+        name: currentVersionCheck
+      - script: echo "##vso[task.setvariable variable=NewVersion;isoutput=true]$(node getCurrentPackageJsonVersion.js)"
+        name: newVersionCheck
+  - job: WaitForNPMPublishValidation
+    displayName: Wait for publishing npm package validation
+    dependsOn: [GetNpmPackageVersions, Build]
+    pool: server
+    variables: 
+      currentVersion: $[dependencies.GetNpmPackageVersions.outputs['currentVersionCheck.CurrentVersion']]
+      newVersion: $[dependencies.GetNpmPackageVersions.outputs['newVersionCheck.NewVersion']]
+    condition: and(succeeded(), ne(variables['Build.Reason'], 'PullRequest'), gt(variables['newVersion'], variables['currentVersion']))
+    steps:
+      - task: ManualValidation@0
+        inputs:
+          notifyUsers: embeddeddev@microsoft.com
+          instructions: "Please validate the vscode-serial-monitor-api npm package release. After validation, resume the pipeline."
+  - job: Publish
+    displayName: Publish the npm package
+    dependsOn: [WaitForNPMPublishValidation, GetNpmPackageVersions]
     pool:
       name: VSEngSS-MicroBuild2022-1ES
+    variables:
+      # MicroBuild requires TeamName to be set.
+      TeamName: C++ Cross Platform and Cloud
+      currentVersion: $[dependencies.GetNpmPackageVersions.outputs['currentVersionCheck.CurrentVersion']]
+      newVersion: $[dependencies.GetNpmPackageVersions.outputs['newVersionCheck.NewVersion']]
+    condition: and(succeeded(), ne(variables['Build.Reason'], 'PullRequest'), gt(variables['newVersion'], variables['currentVersion']))
     steps:
-      - task: UseNode@1
-        displayName: Use Node 16.18
-        inputs:
-          version: "16.18"
-      - script: npm ci
-      - script: npm run build
-      - script: npm publish
-        condition: and(succeeded(), ne(variables['Build.Reason'], 'PullRequest'))
+      - checkout: self
```

Comment:
What does "self" mean here and how is it different than the default behavior?

Reply:
I believe it has to do with "syncing sources". Basically, I believe that this in combination with the persist credentials is what is used to tell the pipeline that we need to be able to modify the repo or at least run git scripts on it. 

https://learn.microsoft.com/en-us/azure/devops/pipelines/yaml-schema/steps-checkout?view=azure-pipelines#examples

Reply:
Thanks for clarifying.